service/history/service.go

```diff
@@ -92,7 +94,19 @@ func (s *Service) Start() {
 
 	historyservice.RegisterHistoryServiceServer(s.server, s.handler)
 	healthpb.RegisterHealthServer(s.server, s.healthServer)
-	s.healthServer.SetServingStatus(serviceName, healthpb.HealthCheckResponse_SERVING)
+
+	// start as NOT_SERVING, update to SERVING after initial shards acquired
+	s.healthServer.SetServingStatus(serviceName, healthpb.HealthCheckResponse_NOT_SERVING)
+	readinessCtx, readinessCancel := context.WithCancel(context.Background())
+	s.readinessCancel = readinessCancel
+	go func() {
+		if s.handler.controller.InitialShardsAcquired(readinessCtx) == nil {
```

Comment:
what should happen if InitialShardsAcquired returns an error or context is canceled? It won't try again. Is that expected?

Reply:
It doesn't return any error besides a context error. And the context is only canceled by Service.Stop, in which case I want this to abort (otherwise it could race and set status back to SERVING after Stop sets it to NOT_SERVING)